Add extra load file tooltip on Linux aimed at IDAaaS users

jclarkeSTFC opened the pull request:
In the new year (2024), the way that users access the archive on ISAaaS will be changing slightly, so they'll need to mount the archive themselves using a new shortcut in the Applications menu. This extra addition to the tooltip will hopefully help them out a bit.

### Description of work
- When searching for a file path, instead of just returning a string (or an empty string for a fail), a ``Result<std::string>`` is returned. This new type wraps the result (if there is one) with any error messages.
- In ``ISISDataArchive``, if it manages to get the path by querying ``data.isis.rl.ac.uk``, but then can't actually get to that file on your machine, then you'll see the new message about IDAaaS.

Fixes #36566. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

### To test:
- New tooltip should only appear on Linux with the archive option set to ``on`` or ``ISIS``.
- Try and load a file that's in the archive, but with the archive not mounted, the new message should appear.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Add extra help message on Linux.

In the new year, the way that users access the archive on ISAaaS will be
changing slightly, so they'll need to mount the archive themselves using
a new shortcut in the Applications menu. This extra addition to the tooltip
will hopefully help them out a bit.

jclarkeSTFC committed:
Added release notes

jclarkeSTFC committed:
Pass errors from the archive search back up

Localise IDAaaS error message to the ISISDataArchive, and pass the error
message back up to anywhere that's trying to load files. Potential there
for adding even more detail in other cases if required.

jclarkeSTFC committed:
Use function overload instead of default arg

Instead of giving the error string a default value in the function
definition (gcc doesn't like that), use an overload instead, then one
can just call the other.

jclarkeSTFC committed:
Use function overload instead of default arg

Some more replacing default error string args with an overload instead.
Had to reorder the arguments in one instance because there were other
arguments with default values, so the error string (now with no default)
had to go first. Hopefully gcc will now be appeased.

jclarkeSTFC committed:
Update cppcheck suppressions

martyngigg reviewed (commented):
This should give ISIS users a hint of why an archive search might not work. I like that the ISIS-specific code has been isolated to the ISIS-specific class.

I've made a suggestion for what I think is a slightly cleaner API but it is a slightly wider change so may be more risk than is acceptable at this point. Although this is the point where the most amount of testing will be done so could be a good time too...

jclarkeSTFC committed:
Use new Result class for archive search

Instead of just returning a string when trying to find a file, return an
object that wraps the string with an error message. Then we can query the
object as to whether it was a success or not instead of just checking
if the returned string is empty. This also allows us to pass messages
around without using input reference parameters.

martyngigg commented on Framework/API/inc/MantidAPI/Result.h:
```diff
@@ -0,0 +1,30 @@
+
+template <typename T> class MANTID_API_DLL Result {
+public:
+  Result(T result, std::string errors = "") : m_result(result), m_errors(errors) {}
```
As there is a copy by the constructor you should be able to `move` the result and errors into the member variables:

```c++
 Result(T result, std::string errors = "") : m_result(std::move(result)), m_errors(std::move(errors)) {}
```

martyngigg commented on Framework/API/inc/MantidAPI/Result.h:
```diff
@@ -0,0 +1,30 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
```
Could we add some tests for this?

martyngigg reviewed (commented):
I think this looks good now. It's good to have an improved api for findRun that better expresses error handling.

A couple of things commented on inline. The only other thing is whether the `Result` type would be better in `Kernel` where it could be used more widely as it's very generic.

jclarkeSTFC committed:
Use move in the constructor

The constructor copies these two objects already, so just move them into
the member variables.

jclarkeSTFC committed:
Update cppcheck suppression

jclarkeSTFC committed:
Remove macro from templated class

jclarkeSTFC committed:
Add tests for new Result class

martyngigg commented on buildconfig/CMake/CppCheck_Suppressions.txt.in:
```diff
@@ -319,7 +319,7 @@ constVariablePointer:${CMAKE_SOURCE_DIR}/Framework/API/src/CompositeFunction.cpp
 variableScope:${CMAKE_SOURCE_DIR}/Framework/API/src/FileProperty.cpp:232
 useStlAlgorithm:${CMAKE_SOURCE_DIR}/Framework/API/src/FileProperty.cpp:296
-constVariableReference:${CMAKE_SOURCE_DIR}/Framework/API/src/FileFinder.cpp:719
+constVariableReference:${CMAKE_SOURCE_DIR}/Framework/API/src/FileFinder.cpp:712
```
While we're changing that code is there any chance we could fix this instead of moving the suppression?

jclarkeSTFC committed:
Fix missing const from cppcheck

jclarkeSTFC committed:
Fix a few accidental string casts

jclarkeSTFC committed:
Propagate error message through FileFinder

martyngigg commented on Framework/API/test/ResultTest.h:
```diff
@@ -0,0 +1,49 @@
+  }
+
+private:
+  template <typename T> void testWithT(T &in) {
```
Just one small comment here. It's good practice when passing by reference to first make things `const T &` as it signals to reader that I don't have to read the code to know if `in` is modified. Also, if you do accidentally modify it then the compiler will shout at you :)

martyngigg commented on Framework/API/test/ResultTest.h:
```diff
@@ -0,0 +1,49 @@
+
+class ResultTest : public CxxTest::TestSuite {
+public:
+  void testStringErrors() {
```
I would also be tempted to split out the tests into those that test success paths and those that test the error paths and consider [naming](https://kevlinhenney.medium.com/a-test-of-knowledge-78f4688dc9cb) them a little more verbosely to indicate what it means if the tests fail.

I'm happy with the current changes if this needs to be done for a code freeze or release window and this change can be done afterward.

jclarkeSTFC committed:
Split tests up into success/failure

Also add a missing const

martyngigg commented:
The code looks good to me and all of the tests pass. I unfortunately don't have easy access to a Linux instance to build and test this. Would the second reviewer be able to do that?

martyngigg commented:
I've chosen not to mark this as approved so that a functional test can be run first as I don't have access to an environment to do that.

cailafinn reviewed (changes requested):
Functional testing: I seem to be getting the error message in there twice. Does this need to be fixed?
<img width="2043" alt="image" src="https://github.com/mantidproject/mantid/assets/47181718/ba87f0cf-79c7-4e2f-bdca-0baf2437cd46">

jclarkeSTFC commented:
> Functional testing: I seem to be getting the error message in there twice. Does this need to be fixed? <img alt="image" width="2000" src="https://private-user-images.githubusercontent.com/47181718/295231668-ba87f0cf-79c7-4e2f-bdca-0baf2437cd46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MTAyNTEsIm5iZiI6MTcwNDgwOTk1MSwicGF0aCI6Ii80NzE4MTcxOC8yOTUyMzE2NjgtYmE4N2YwY2YtNzljNy00ZTJmLWJkY2EtMGJhZjI0MzdjZDQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDE0MTkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0YmJhYTY3YjA2NDk3NTIxZjMxYjFhNjM3ZWFiY2RjNDgxZTI1NmUyMGJkMWM5NzMwY2YxNmVmYzcwNTFkNDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._0RNg-thG5KaUxcFHGdMSqswhU4FDtQweY7XTkK3Mh0">

I'll have a look, probably once is enough

jclarkeSTFC commented:
> Functional testing: I seem to be getting the error message in there twice. Does this need to be fixed? <img alt="image" width="2000" src="https://private-user-images.githubusercontent.com/47181718/295231668-ba87f0cf-79c7-4e2f-bdca-0baf2437cd46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MTEyMTcsIm5iZiI6MTcwNDgxMDkxNywicGF0aCI6Ii80NzE4MTcxOC8yOTUyMzE2NjgtYmE4N2YwY2YtNzljNy00ZTJmLWJkY2EtMGJhZjI0MzdjZDQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDE0MzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3NTFhZGI0ZmQxMmI4MGY3MzYxYzk5NDE0ZTJkYjMxZDRkOWM4ZjVlODkwNDYwMmFhNzg0YjU0Y2IxNmRiMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.38BYyjF2MBDsto7pqnmjQ09W_yxuslJGxtYjQO6qUf0">

The message only appears once for me, are there some other steps to getting the extra text?

cailafinn commented:
> The message only appears once for me, are there some other steps to getting the extra text?

I just followed the normal steps. Unmounted the archive and then attempted to load a file that would be in it. It's on a release build, that could make a difference?

jclarkeSTFC committed:
Merge branch 'main' into 36566_idaaas_archive_tooltip

jclarkeSTFC committed:
Remove superfluous suppression

This line should have been deleted in the merge.

cailafinn reviewed (approved):
Given it seems to just be me with the double text issue I’m happy to approve this.
